### PR TITLE
[CLIPPER-65] Performance / Resilience Test Based on CIFAR Demo

### PR DESCRIPTION
**Pull request by Corey-Zumar**

### CODE REVIEW

**Comment by Corey-Zumar** [comment]
@dcrankshaw

**Comment by dcrankshaw** [comment]
jenkins test this please

File: bench/bench_init.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+			test_x, test_y, train_x, train_y = self.get_cifar_data()
+			print("Training model...")
+			model = lm.LogisticRegression()
+			model.fit(train_x, train_y)
+			joblib.dump(model, model_location)
+			print("Model trained!")
+			print("Logistic Regression test score: %f" % model.score(test_x, test_y))
+
+if __name__ == '__main__':
+	setup = BenchSetup("localhost", "../examples/cifar_demo")
```
Comment (dcrankshaw) [inline bench/bench_init.py]: Can you have the location of the CIFAR dataset be a command line argument?

File: bench/bench_init.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	def get_cifar_data(self):
+		train_x, train_y = cifar_utils.filter_data(
+			*cifar_utils.load_cifar(self.cifar_dir_path, cifar_filename="cifar_train.data", norm=False))
+		test_x, test_y = cifar_utils.filter_data(
+			*cifar_utils.load_cifar(self.cifar_dir_path, cifar_filename="cifar_test.data", norm=False))
+
+		return test_x, test_y, train_x, train_y
+
+	def train_sklearn_model(self):
+		model_location = BASE_DATA_PATH + SKLEARN_MODEL_FILE
```
Comment (dcrankshaw) [inline bench/bench_init.py]: use `os.path.join()` not string concatenation

File: bench/bench_init.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import sys
+import os
+import json
+import requests
```
Comment (dcrankshaw) [inline bench/bench_init.py]: json and requests imports aren't used

File: bench/bench_init.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+			model = lm.LogisticRegression()
+			model.fit(train_x, train_y)
+			joblib.dump(model, model_location)
+			print("Model trained!")
+			print("Logistic Regression test score: %f" % model.score(test_x, test_y))
+
+if __name__ == '__main__':
+	setup = BenchSetup("localhost", "../examples/cifar_demo")
+	setup.run()
+
```
Comment (dcrankshaw) [inline bench/bench_init.py]: Get rid of the tabs in this file

**Comment by dcrankshaw** [comment]
jenkins, test this please

**Comment by dcrankshaw** [comment]
@Corey-Zumar Looks like there's a problem with Jenkins, don't worry about the failing tests for now. I'm reviewing this right now.

**Comment by shaneknapp** [comment]
test this please

**Comment by shaneknapp** [comment]
test this please

**Comment by shaneknapp** [comment]
test this please

File: bench/bench_init.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import json
+import requests
+sys.path.append("..")
+
+from management import clipper_manager
+from examples.tutorial import cifar_utils
+from sklearn import linear_model as lm
+from sklearn.externals import joblib
+from fabric.api import *
+
```
Comment (dcrankshaw) [inline bench/bench_init.py]: In order to get the imports to work correctly, I added the following files
```
management/__init__.py
examples/__init__.py
examples/tutorial/__init__.py
```

Then I changed the imports to the following:
```py
sys.path.append(os.path.abspath("../management"))
sys.path.append(os.path.abspath("../examples"))

import clipper_manager
from tutorial import cifar_utils
from sklearn import linear_model as lm
from sklearn.externals import joblib
from fabric.api import *
```

Comment (dcrankshaw) [reply bench/bench_init.py]: We will want to rethink directory organization in the future, but for now this should work. I don't want an __init__.py file in the root of the source code.

File: bench/bench_init.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+APP_NAME = "bench"
+BASE_DATA_PATH = "data/"
```
Comment (dcrankshaw) [inline bench/bench_init.py]: This is confusingly named, because this is where the model, not the data, will be stored.

Change it to:

```py
BASE_MODEL_PATH = model/"
```

File: bench/bench_init.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+		if os.path.isfile(model_location):
+			model = joblib.load(model_location)
+			print("Found and loaded model!")
+		else:
+			print("Loading CIFAR data...")
+			test_x, test_y, train_x, train_y = self.get_cifar_data()
+			print("Training model...")
+			model = lm.LogisticRegression()
+			model.fit(train_x, train_y)
+			joblib.dump(model, model_location)
```
Comment (dcrankshaw) [inline bench/bench_init.py]: This will fail if `BASE_DATA_PATH` doesn't exist yet. Add a check to see if the directory exists, and if it doesn't, create it.

File: src/frontends/src/bench.cpp [hunk 8]
```diff
@@ -3,21 +3,159 @@
...
+  std::cout << "Enter the delay between batches, in milliseconds: ";
+  std::cin >> batch_delay;
+
+  config.emplace(CONFIG_KEY_PATH, path);
+  config.emplace(CONFIG_KEY_NUM_THREADS, num_threads);
+  config.emplace(CONFIG_KEY_NUM_BATCHES, num_batches);
+  config.emplace(CONFIG_KEY_BATCH_SIZE, batch_size);
+  config.emplace(CONFIG_KEY_BATCH_DELAY, batch_delay);
+  return config;
+};
```
Comment (dcrankshaw) [inline src/frontends/src/bench.cpp:161]: Reading from the command-line is really nice for one-off benchmarks but impossible to script.

Can you a command line arg `-f FILENAME.json` that reads a JSON config file to get the benchmark parameters. We already have a command line options library in the codebase, check out [`query_frontend_main.cpp`](https://github.com/ucbrise/clipper/blob/35cb8ee0a9fe8f8ea8601d06b8cc8adcfe2fc227/src/frontends/src/query_frontend_main.cpp#L2-L18) for an example on how to use it.

The behavior should be to first parse command line args. If the filename arg is present, read the JSON file to get the benchmark params. If it's not present, use the existing approach of asking the user interactively.

File: src/frontends/src/bench.cpp [hunk 9]
```diff
@@ -3,21 +3,159 @@
...
+constexpr double SKLEARN_BIRD_LABEL = 0;
+
+std::unordered_map<int, std::vector<std::vector<double>>> load_cifar(
+    std::unordered_map<std::string, std::string>& config) {
+  std::ifstream cifar_file(config.find(CONFIG_KEY_PATH)->second, std::ios::binary);
+  std::istreambuf_iterator<char> cifar_data(cifar_file);
+  std::unordered_map<int, std::vector<std::vector<double>>> vecs_map;
+  for (int i = 0; i < 10000; i++) {
+    int label = static_cast<int>(*cifar_data);
+    std::cout << label << std::endl;
```
Comment (dcrankshaw) [inline src/frontends/src/bench.cpp]: Don't print the label here.

File: src/frontends/src/bench.cpp [hunk 10]
```diff
@@ -3,21 +3,159 @@
...
+      vecs_map.emplace(label, new_label_vecs);
+    }
+  }
+  return vecs_map;
+}
+
+void send_predictions(std::unordered_map<std::string, std::string>& config, QueryProcessor &qp,
+                      std::unordered_map<int, std::vector<std::vector<double>>> &cifar_data,
+                      std::shared_ptr<metrics::RatioCounter> accuracy_ratio) {
+  std::vector<std::vector<double>> planes_vecs = cifar_data.find(0)->second;
```
Comment (dcrankshaw) [inline src/frontends/src/bench.cpp]: I get a segfault at this line. I'm guessing something is wrong with loading the cifar data?

Let me know if you can't reproduce it, and I can dig into it a little more.

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
This is a good start. There's a bug in the benchmark code though, so I haven't been able to actually get through the full benchmark.

Also, can you format and lint the Python code to make sure it conforms with PEP8?

File: bench/bench_init.py (hunk 5 above)
Comment (Corey-Zumar) [reply bench/bench_init.py]: Replaced imports with the one's you've defined.

File: bench/bench_init.py (hunk 3 above)
Comment (Corey-Zumar) [reply bench/bench_init.py]: Removed

File: bench/bench_init.py (hunk 6 above)
Comment (Corey-Zumar) [reply bench/bench_init.py]: Agreed. Changed this!

File: bench/bench_init.py (hunk 2 above)
Comment (Corey-Zumar) [reply bench/bench_init.py]: Done

File: bench/bench_init.py (hunk 7 above)
Comment (Corey-Zumar) [reply bench/bench_init.py]: Done.

File: bench/bench_init.py (hunk 1 above)
Comment (Corey-Zumar) [reply bench/bench_init.py]: Done. Also modified `setup_bench.sh` to accept this path as a command line argument.

File: src/frontends/src/bench.cpp (hunk 10 above)
Comment (Corey-Zumar) [reply src/frontends/src/bench.cpp]: I think the instructions I provided for running the benchmark are insufficient / unclear. For the path to the CIFAR binary data set, I'm passing a path to a file within the **binary** version of CIFAR hosted at https://www.cs.toronto.edu/~kriz/cifar.html . I'm using `data_batch_1.bin`. I should have stated this explicitly.

If you were passing a path to analogous file (`data_batch_<n>.bin`) within the binary data set and still received the seg fault, please let me know.

File: src/frontends/src/bench.cpp (hunk 8 above)
Comment (Corey-Zumar) [reply src/frontends/src/bench.cpp:161]: Done. We can now specify a path to a JSON configuration file with the command line argument `-f ` or `--filename`. The configuration file should be formatted as follows:

```
{
   "cifar_data_path":"<path to a data_batch_<n>.bin file in the cifar binary data set>",
   "num_threads":"<number of threads>",
   "num_batches":"<number of batches>",
   "batch_size":"<batch size>",
   "batch_delay_millis":"<batch delay millis>"
}
```

**Comment by dcrankshaw** [comment]
@Corey-Zumar This is reviewed and ready for you to address the comments

File: src/frontends/src/bench.cpp (hunk 9 above)
Comment (Corey-Zumar) [reply src/frontends/src/bench.cpp]: Removed this line

**Comment by dcrankshaw** [comment]
jenkins test this please

**Comment by Corey-Zumar** [comment]
Formatted `bench_init.py` to conform to PEP8

**Comment by dcrankshaw** [comment]
jenkins test this please

File: bench/setup_bench.sh [hunk 11]
```diff
@@ -0,0 +1,13 @@
+trap "exit" INT TERM
+trap "kill 0" EXIT
+
+echo -n "Enter path to the CIFAR python data set: "
+read cifar_path
```
Comment (dcrankshaw) [inline bench/setup_bench.sh]: Don't interactively ask for input, just take the location as a command line param:

```
cifar_path=$1
```

File: src/frontends/src/bench.cpp [hunk 12]
```diff
@@ -3,21 +3,199 @@
...
+      send_predictions(test_config, qp, cifar_data, accuracy_ratio);
+    });
+    threads.push_back(std::move(thread));
+  }
+  for (auto &thread : threads) {
+    thread.join();
+  }
+  std::string metrics = metrics::MetricsRegistry::get_metrics().report_metrics();
+  log_info("BENCH", metrics);
+  std::this_thread::sleep_for(std::chrono::seconds(5));
```
Comment (dcrankshaw) [inline src/frontends/src/bench.cpp]: Why are you sleeping at the end here?

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
Looks good. A couple minor comments.

Can you add some documentation about how to run the benchmark? At a minimum, you need to specify what the dataset for the model container should be and what the dataset for the bench binary should be.

Also, I found a bug in the RPCService destructor. 


Replace [`RPCService::shutdown_service`](https://github.com/ucbrise/clipper/blob/35cb8ee0a9fe8f8ea8601d06b8cc8adcfe2fc227/src/libclipper/src/rpc_service.cpp#L130-L133) with
```cpp
void RPCService::shutdown_service(socket_t &socket) {
  size_t buf_size = 32;
  char buf[buf_size];
  socket.getsockopt(ZMQ_LAST_ENDPOINT, (void *)&buf, &buf_size);
  std::string last_endpoint = std::string(buf);
  socket.unbind(last_endpoint);
  socket.close();
 }
```

The problem is we were calling `disconnect` not `unbind`. The mess with ZMQ_LAST_ENDPOINT is because we are using wildcard addresses (http://api.zeromq.org/4-2:zmq-unbind#toc3).

**Comment by dcrankshaw** [comment]
jenkins test this please

File: bench/setup_bench.sh (hunk 11 above)
Comment (Corey-Zumar) [reply bench/setup_bench.sh]: Done

File: src/frontends/src/bench.cpp (hunk 12 above)
Comment (Corey-Zumar) [reply src/frontends/src/bench.cpp]: I shouldn't be. I think this was an illogical attempt to log metrics in the face of the RPC error you identified above. Removed.

**Review by Corey-Zumar — COMMENTED** [review]
Added a README with a list of required files and a detailed explanation of how to run the test. I also integrated your fix for the RPC error that you identified - thanks for finding that - and addressed your other comments.

File: bench/README.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
+# Running the performance benchmark
+
+## Required Files
+### CIFAR10 Python Dataset for SKLearn Model
+This benchmark serves an SKLearn model that depends on the CIFAR10 Python dataset for training. This dataset can be obtained from [https://www.cs.toronto.edu/~kriz/cifar.html](https://www.cs.toronto.edu/~kriz/cifar.html).
```
Comment (dcrankshaw) [inline bench/README.md]: This actually isn't true. The `cifar_utils` functions expect a CSV. You can use https://github.com/ucbrise/clipper/blob/develop/examples/tutorial/download_cifar.py to download the dataset and transform it into a CSV. If you look at the [tutorial](https://github.com/ucbrise/clipper/blob/develop/examples/tutorial/tutorial_part_one.ipynb), we first download the dataset and create the CSV, then load it to train the model.

**Review by dcrankshaw — COMMENTED** [review]
Looks good other than some confusion in the docs about the CIFAR python dataset.

File: bench/README.md (hunk 13 above)
Comment (Corey-Zumar) [reply bench/README.md]: Thanks! Fixed this. Please take another look.

**Review by dcrankshaw — APPROVED** [review]

**Comment by dcrankshaw** [comment]
jenkins test this please
